File: recursion/python/solutions/split_at.py
"""Solution for split_at."""

from linked_list import EMPTY_LIST, Node, is_empty, prepend, unprepend, to_python_list


def split_at(n, lst):
    """
    Split list at index n into (first n elements, rest).

    split_at(2, 1 -> 2 -> 3 -> 4) => (1 -> 2, 3 -> 4)
    split_at(0, 1 -> 2 -> 3) => ((), 1 -> 2 -> 3)
    """
    if n == 0 or is_empty(lst):
        return (EMPTY_LIST, lst)

    first, rest = unprepend(lst)
    left, right = split_at(n - 1, rest)

    return (prepend(first, left), right)


if __name__ == '__main__':
    # Test cases
    left, right = split_at(2, Node.from_list([1, 2, 3, 4]))
    assert to_python_list(left) == [1, 2]
    assert to_python_list(right) == [3, 4]

    left, right = split_at(0, Node.from_list([1, 2, 3]))
    assert to_python_list(left) == []
    assert to_python_list(right) == [1, 2, 3]

    left, right = split_at(5, Node.from_list([1, 2]))
    assert to_python_list(left) == [1, 2]
    assert to_python_list(right) == []

    print("All tests passed!")

File: recursion/javascript/solutions/splitAt.js
/**
 * Solution for splitAt.
 */

const { EMPTY_LIST, Node, isEmpty, prepend, unprepend, toArray } = require('./linkedList');

function splitAt(n, list) {
  if (n === 0 || isEmpty(list)) {
    return [EMPTY_LIST, list];
  }

  let [first, rest] = unprepend(list);
  let [left, right] = splitAt(n - 1, rest);

  return [prepend(first, left), right];
}

if (require.main === module) {
  let [left, right] = splitAt(2, Node.fromArray([1, 2, 3, 4]));
  console.assert(JSON.stringify(toArray(left)) === '[1,2]');
  console.assert(JSON.stringify(toArray(right)) === '[3,4]');

  let [l2, r2] = splitAt(0, Node.fromArray([1, 2, 3]));
  console.assert(JSON.stringify(toArray(l2)) === '[]');
  console.assert(JSON.stringify(toArray(r2)) === '[1,2,3]');

  let [l3, r3] = splitAt(5, Node.fromArray([1, 2]));
  console.assert(JSON.stringify(toArray(l3)) === '[1,2]');
  console.assert(JSON.stringify(toArray(r3)) === '[]');

  console.log('All tests passed!');
}

module.exports = { splitAt };
